### ISSUE TITLE
Allow simple customization of sign-up and sign-in attributes in React Authenticator

### ISSUE THREAD

**Issue by kennu** [aws-amplify/amplify-js#2160]
It is currently a very heavy operation to customize the default SignIn and SignUp views of the React Authenticator component. All I wanted was to get rid of unnecessary fields to have simple Email + Password login. To achieve it, I had to clone the SignUp and SignIn components and half-rewrite them with the appropriate form fields.

I think developers should be able to specify the sign-up and sign-in attributes as a simple list like ['email', 'password'], and the React components would adapt automatically based on it.

As an alternative, at least please provide some kind of a preset or default for Email + Password login. I think it's the most popular way to sign up nowadays. Phone number and custom username should be optional extras.

**Comment by haverchuck** [comment]
@kennu We have an implementation of this on our @beta tag.  Simply do an `npm install aws-amplify-react@beta` to try it out.  Keep in mind that, as the name suggests, this is in beta.

If you are using the Authenticator component without the HOC, you can pass custom fields like so:

`  <Authenticator signUpConfig={signUpConfig}/>`

If you are using the HOC, you can pass it like so:

`export default withAuthenticator(App, { signUpConfig });`

The signUpConfig object has the following schema:

![image](https://user-images.githubusercontent.com/10621276/48861847-73893e00-ed79-11e8-94d0-cb1f707d8f54.png)

The signUpField objects are in turn structured this way:
![image](https://user-images.githubusercontent.com/10621276/48861898-94ea2a00-ed79-11e8-82a8-381408114622.png)

In short, this functionality should allow you to pass in your own fields, hide all or some of the defaults, and set the default countryCode in the phone number code dropdown.

Please feel free to let us know what you think.

**Comment by kennu** [comment]
Thanks @haverchuck it looks really good. Assuming that you have made the basic Email+Password sign-up case work, I think it will cover the main need I have.

**Comment by haverchuck** [comment]
@kennu I don't think this will actually use email as the username by default.... maybe you could make that work by passing in a field as... 
```
{
   key: 'username',
   label: 'email'
}
```

I'm not positive that would work - it might be tricky to do your own validation before the SignUp component sends the request.  I'm thinking we should probably implement a callback.

**Comment by kennu** [comment]
If you mean to use email as Username instead of an UUID, then I don't actually want that. It's good that the permanent Username is a random UUID generated by Cognito, so that the email can be changed later. I just want the user to be able to sign in by only giving their email and not have to choose any username manually.

**Comment by mrcoles** [comment]
I’m looking at this right now, trying to get a sign up form that shows just email + password. Has anyone gotten this to actually work with email + password? I have run into the following issues:

**1\. Basic attempt**

I tried a very simple starting approach with a signUpConfig as such:

```javascript
const signUpConfig = {
  signUpFields: [ { key: 'email' }, { key: 'password' } ]
}
```

This still rendered the full form, but the email and password widgets lost all of their labels.

Also, password was ordered before email.

Suggestion: it would be cool if the custom signUp fields extended the defaults, e.g., `field = Object.assign({}, defaultEmailField, customEmailField)`.

**2\. Full setup**

I used `hideAllDefaults` to exclude the fields I did not want.

I dug through the source code and found how the default fields were structured: [default-sign-in-fields.js](https://github.com/aws-amplify/amplify-js/blob/44cf04975ba41d08581480ff3dd79476b2ff6477/packages/aws-amplify-react-native/src/Auth/common/default-sign-in-fields.js)

So, I then tried with this approach:

```javascript
const signUpConfig = {
  hideAllDefaults: true,
  signUpFields: [
    {
      label: 'Email',
      key: 'email',
      required: true,
      placeholder: 'Email',
      type: 'email',
      displayOrder: 1
    },
    {
      label: 'Password',
      key: 'password',
      required: true,
      placeholder: 'Password',
      type: 'password',
      displayOrder: 2
    }
  ]
};
```

This appeared to render correctly, but when I entered an email and password and submitted the form, I got the following error:

> Uncaught TypeError: Cannot read property 'replace' of undefined  
>     at SignUp.jsx:191  
>     at Array.forEach (<anonymous>)  
>     at SignUp.signUp (SignUp.jsx:185)

Which is [this line of code](https://github.com/aws-amplify/amplify-js/blob/dbc4c01b2f05e05565fe0dd0de9625059bcbde75/packages/aws-amplify-react/src/Auth/SignUp.jsx#L191).

Jumping into the debugger, I see the following values for these local variables:

```javascript
inputKeys = ['username', 'checkedValue', 'password', 'dial_code'];
inputVals = ['test@example.com', null, 'Password', '+1'];
signup_info: { attributes: {}, password: 'Password', username: 'test@example.com' };
```

It looks like [this line of code](https://github.com/aws-amplify/amplify-js/blob/dbc4c01b2f05e05565fe0dd0de9625059bcbde75/packages/aws-amplify-react/src/Auth/SignUp.jsx#L163) is to blame:

```javascript
        if (!this.inputs.dial_code) {
            this.inputs.dial_code = this.getDefaultDialCode();
        }
```

Either this needs to not happen when there’s no `phone_line_number` field or the code that combines them to make `phone_number` needs to ignore it if there is no `phone_line_number`.

**3\. Aside about `displayOrder`**

I originally had email with `displayOrder: 0` and password with `displayOrder: 1`, this resulted in password showing up first. I think [this line in SignUp.jsx](https://github.com/aws-amplify/amplify-js/blob/44cf04975ba41d08581480ff3dd79476b2ff6477/packages/aws-amplify-react/src/Auth/SignUp.jsx#L215) (and the one in aws-amplify-reactnative too) should actually be:

```javascript
if (typeof(a.displayOrder) === 'number' && typeof(b.displayOrder) === 'number') {
```

**Summary**

Has anyone been able to run this actual example of an email + password SignUp using the `withAuthenticator` HOC or `Authenticator` component?

Bugs/Feature requests:

1. It would be awesome if custom signUpFields would extend the defaults
2. The code forces a dial_code input value even if you don’t include a phone number and this throws an unchecked error
3. The displayOrder value should support `0`
4. As mentioned by @haverchuck, there is no support for overriding SignIn yet either. My assumption is that the `username` field would default to email sign
5. Hitting "return" while inside a field does not submit the form. This is breaks user expectations of how an HTML form should work.

**Comment by mrcoles** [comment]
@haverchuck have there been updates on this code since? I would assume at least the `dial_code` bug was fixed?

**Comment by haverchuck** [comment]
@mrcoles The dial code bug should be fixed in our @unstable build (or rather, it will be fixed shortly once the deployment finishes).

**Comment by josephecombs** [comment]
We need way more code samples of valid configurations of the signUpConfig object. This should with with just email and password, out of the box.

**Comment by jameskhamil** [comment]
@haverchuck is there any chance username can be automatically generated/UUID/email? Currently using one of the solutions on this page produces `Username cannot be empty`.  Username as a concept has limited usage across different types of content.

**Comment by slavagu** [comment]
This snippet worked for me:
```
const signUpConfig = {
  hideAllDefaults: true,
  signUpFields: [
    {
      label: 'Email',
      key: 'username',
      required: true,
      placeholder: 'Email',
      type: 'email',
      displayOrder: 1,
    },
    {
      label: 'Password',
      key: 'password',
      required: true,
      placeholder: 'Password',
      type: 'password',
      displayOrder: 2,
    },
  ],
}

const Login = () => <Authenticator signUpConfig={signUpConfig} />
```

Note the 1st field key is set to `username` instead of `email`, which works as expected when the User Pool is created with 
```
      UsernameAttributes:
        - 'email'
```
so that Users can use an email address or phone number as their "username" to sign up and sign in:
![image](https://user-images.githubusercontent.com/3827291/53692786-d7e08280-3de9-11e9-957a-fd26fb41f707.png)

An interesting (and desired) side effect of signing up the user with email instead of username, is that username becomes a GUID generated by Cognito:
![image](https://user-images.githubusercontent.com/3827291/53692824-84226900-3dea-11e9-887b-3fc939e98b33.png)

Overall, happy with the solution. Thanks folks!

**Comment by mikekhristo** [comment]
I think you’ll also need to force the email value to be lower case.

**Comment by slavagu** [comment]
> I think you’ll also need to force the email value to be lower case.
Just tried with upper case and it worked fine. Sign Up and Sign In both worked for email like  MixedCase@example.com

**Comment by mrcoles** [comment]
@slavagu just to warn you—from my experience with AWS Cognito, it is case-sensitive with emails, so you can have different accounts for both MixedCase@example.com and mixedcase@example.com. 

This is really strange, considering amazon.com itself (along with every other website I can think of) allows case-insensitive email logins.

If you’re using this UI for login, you can update things client-side before they’re sent to the servers, so I’d recommend following @mikekhristo’s advice and lowercase emails ahead of time. NOTE: if you ever switch to the hosted UI, there’s no control over this and the preAuth lambda functions are quite limited and cannot override these values either.

Here are various links to the issue. Unfortunately, I don’t see any way to bring this up directly with the Cognito team, as they don’t seem to have a public-facing issues tracker like amplify-js or amplify-cli do.

*  https://github.com/amazon-archives/amazon-cognito-identity-js/issues/658  
*  https://github.com/aws-amplify/amplify-js/issues/610  
*  https://stackoverflow.com/questions/48342450/aws-cognito-username-email-login-is-case-sensitive  
*  https://www.reddit.com/r/aws/comments/8p5qm6/have_i_found_a_bug_in_cognito/

**Comment by kennu** [comment]
Technically the local-part of email addresses may be case-sensitive and it's decided locally by each email system, so it would be false to assume that email addresses are always case-insensitive. https://tools.ietf.org/html/rfc5321#section-4.1.2

Regardless of that, it would be nice to have more control to override things in the Cognito Lambda event handlers.

**Comment by mrcoles** [comment]
Yeah, you hit the nail on the head. The RFC spec does say that and it’s at odds with how every major website uses emails to identify users, so the technical interpretation of the spec doesn’t service the needs of most people using the product. I agree it’d be really helpful to have an option to override this.

I would prefer this were an option in the Cognito pool itself, so it could be set correctly in one place and less error-prone & easier for setups.

As a consolation, I’d take the DIY Lambda event handlers approach if I could do it in “Pre sign-up” and “Pre authentication”, but I would assume those run too late in the process, so I’d even take some additional new trigger “Transform attributes” or something that runs before those two other triggers?

Also, we’ve gone off topic from the original issue here, sorry everyone!

**Comment by mikhael28** [comment]
The configuration settings for the SignUp Fields work perfectly, except the defaultCountryCode when set to 256 (which is an available preset) doesn't work. In fact, any code doesn't work - it just defaults to 1. I had to go into the node_module and update the preset to 256.

A question - are there currently any ways to update the signIN forms? All these configuration options are for the signUp. I would like to change the signIn form to have the same formatting as the phone_number form in signUp, as my username is a phone number. Am currently modifying the underlying node_module components to see if I can update anything apart from the description.

**Comment by tbehunin** [comment]
I've gotten the front-end to work as designed using the suggestions for `signUpConfig` parameter. The issue now is that **I'm not receiving the email to verify the email address** after signup.

I have the `AutoVerifiedAttributes` value on my Cognito User Pool configured as a list of just `email` specified. When I signup, the user is created in my User Pool as `UNCONFIRMED`, but the email verification isn't sent and I can't seem to find any logs to debug what might be happening. Anyone run into this same issue and have a solution? Disclaimer: I'm using Serverless to deploy my Cognito User Pool and I'm only using Amplify's auth api - not using the Amplify cli to install backend services, etc (since I have Serverless doing all of that for me).

**Comment by ilyador** [comment]
The documentation saying `‘string’, ‘number’, ‘password’, etc` is very lacking since it says nowhere what does `etc` actually include.
Apparently 'date' works, but seems like there are no options for a select dropdown, which is quite a basic option for a signup form.

**Comment by leahjlou** [comment]
@ilyador The "etc" includes any valid value for the `type` attribute on an HTML `<input>`: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input

Agreed that more input types, like a select dropdown, would be really useful.

**Comment by ilyador** [comment]
@leahjlou Thanx, however it does not since `radio` for example, doesn't work.

**Comment by cpmech** [comment]
> The configuration settings for the SignUp Fields work perfectly, except the defaultCountryCode when set to 256 (which is an available preset) doesn't work. In fact, any code doesn't work - it just defaults to 1. I had to go into the node_module and update the preset to 256.
> 
> A question - are there currently any ways to update the signIN forms? All these configuration options are for the signUp. I would like to change the signIn form to have the same formatting as the phone_number form in signUp, as my username is a phone number. Am currently modifying the underlying node_module components to see if I can update anything apart from the description.

Also, options for configuring the Confirm Sign Up form are welcome!

**Comment by kAliRM70** [comment]
Hi All,
Is there a way to add custom drop down select to the withAuthenticator aes-amplify-react UI component; for use example to select user type during sign up ? You support is appreciated

**Comment by elorzafe** [comment]
@kennu @mrcoles from today new User Pools can be created with case insensitivity for username input
More info [here](https://aws.amazon.com/about-aws/whats-new/2020/02/amazon-cognito-user-pools-service-now-supports-case-insensitivity-for-user-aliases)

**Comment by ramanasak** [comment]
how to change background colors, label color text color, etc

**Comment by mrcoles** [comment]
@elorzafe cool! I see this can’t be applied to an existing pool. Do you know where the most updated guide is to migrate from one user pool to another? Is this the best one? https://medium.com/collaborne-engineering/migrate-aws-cognito-user-pools-ff2a91a745a2

**Comment by khanguslee** [comment]
> The documentation saying `‘string’, ‘number’, ‘password’, etc` is very lacking since it says nowhere what does `etc` actually include.
> Apparently 'date' works, but seems like there are no options for a select dropdown, which is quite a basic option for a signup form.

Any update on having the ability to add a select dropdown?

**Comment by thisismydesign** [comment]
@haverchuck Supposed these 2 should be equivalent.

`<Authenticator signUpConfig={signUpConfig}/>`

`export default withAuthenticator(App, { signUpConfig });`

That's not the case for me with TS. The example config provided by @slavagu fails with `withAuthenticator` with a type error while it works with `Authenticator`.

**Comment by sangdth** [comment]
This works for me:
```
withAuthenticator(App, true, [], null, null, signUpConfig));
```

**Comment by Daniyal-star** [comment]
Im facing issues when using custom attribute in signup . My custom attribute is custom:Username .
How im gonna add this in signUpconfigField : [{ }] ?
can anyone help me on this
